### PR TITLE
backupccl: create readAsOfIterator for RESTORE

### LINKED ISSUE #77276: backupccl: create readAsOfMVCCIterator
Currently, while ingesting SSTs in the restore processor, `processRestoreSpanEntry` in `restore_data_processor.go` manually skips delete tombstones and keys with timestamps later than the AS OF SYSTEM TIME timestamp provided in the RESTORE command. 

This logic should be abstracted away from the restore processor. Specifically, when iterating through an SST, `processRestoreSpanEntry` should use a `readAsOfMVCCIterator`, an iterator  that wraps a `SimpleMVCCIterator` such that all keys returned are not tombstones and will have timestamps less than the AOST timestamp,

Related to #71155

Epic [CRDB-11252](https://cockroachlabs.atlassian.net/browse/CRDB-11252)

Jira issue: CRDB-13513

### PR DESCRIPTION
**Pull request by msbutler**
Previously while ingesting SSTs in the restore processor,
`processRestoreSpanEntry` would manually skip delete tombstones and keys with
timestamps later than the AS OF SYSTEM TIME timestamp provided in the RESTORE
command. This PR wraps the restore processor's iterator with a new
readAsOfIterator which abstracts this logic away.

The readAsOfIterator wraps a SimpleMVCCIterator
ond only surfaces the latest valid key of a given MVCC key that is also below
the iterator's AOST timestamp, if set. Further, the iterator does not surface
delete tombstones, nor any MVCC keys shadowed by delete tombstones below the
AOST timestamp, if set.

Fixes https://github.com/cockroachdb/cockroach/issues/77276

Release note: None

### CODE REVIEW

File: pkg/storage/read_as_of_iterator.go
Comment (erikgrinaker) [inline pkg/storage/read_as_of_iterator.go:45]: nit: for most interface methods that don't do anything out of the ordinary, it's better to just say "Valid implements SimpleMVCCIterator." so we don't have to update a ton of comments if we ever change anything.

File: pkg/storage/read_as_of_iterator.go
Comment (erikgrinaker) [inline pkg/storage/read_as_of_iterator.go:23]: nit: I'd call this `asOf`, since this is the timestamp that the iterator reads as of.

File: pkg/storage/read_as_of_iterator.go
Comment (erikgrinaker) [inline pkg/storage/read_as_of_iterator.go:26]: I don't see why we need this field. All callers call `Valid()`, which sets this field as a side-effect, and then they read this field. Why don't the callers just read the return value of `Valid()` directly instead?

File: pkg/storage/read_as_of_iterator.go
Comment (erikgrinaker) [inline pkg/storage/read_as_of_iterator.go:107]: I don't think this is correct: `NextKey()` could land on a non-tombstone key that's above `endTime`.

File: pkg/storage/read_as_of_iterator.go
Comment (erikgrinaker) [inline pkg/storage/read_as_of_iterator.go:119]: nit: we usually call these `New`, not `Make`.

**Review by erikgrinaker — COMMENTED** [review]
Approach looks good, but it could use more tests for corner cases.

File: pkg/storage/read_as_of_iterator.go
Comment (dt) [reply pkg/storage/read_as_of_iterator.go:119]: (more specifically, the distinction is usually if it has a pointer return type or not)

File: pkg/storage/read_as_of_iterator.go
Comment (msbutler) [reply pkg/storage/read_as_of_iterator.go]: Done.

File: pkg/storage/read_as_of_iterator.go
Comment (msbutler) [reply pkg/storage/read_as_of_iterator.go]: Done.

File: pkg/storage/read_as_of_iterator.go
Comment (msbutler) [reply pkg/storage/read_as_of_iterator.go:127]: Good catch. Completely rewrote this.

File: pkg/storage/read_as_of_iterator.go
Comment (msbutler) [reply pkg/storage/read_as_of_iterator.go]: Done.

File: pkg/storage/read_as_of_iterator.go
Comment (msbutler) [reply pkg/storage/read_as_of_iterator.go]: Done.

File: pkg/storage/read_as_of_iterator.go
Comment (erikgrinaker) [inline pkg/storage/read_as_of_iterator.go:91]: This function can be simplified as:

```go
func (f *ReadAsOfIterator) advance() bool {
	for {
		if ok, err := f.Valid(); err != nil || !ok {
			// No valid keys found.
			return false
		} else if !f.asOf.IsEmpty() && f.asOf.Less(f.iter.UnsafeKey().Timestamp) {
			// Skip keys above the asOf timestamp.
			f.iter.Next()
		} else if len(f.iter.UnsafeValue()) == 0 {
			// Skip to the next key if we find a tombstone.
			f.iter.NextKey()
		} else {
			// On a valid key.
			return true
		}
	}
}
```

File: pkg/storage/read_as_of_iterator.go
Comment (msbutler) [reply pkg/storage/read_as_of_iterator.go:15]: Before I respond to the other feedback, we should get on the same page about this. Here's my intention:
- this iterator can surface all versions of a key below a timestamp, via the `Next` function. In `TestReadAsOfIterator`, the following test: 
    ```
    // ensure next key captures at most one mvcc key per key after an asOf skip
    {input: "b3c2c1", expectedNextKey: "c2", expectedNext: "c2c1", asOf: "2"},
    ```

- ...as long as those keys are _not_ shadowed by a tombstone below the asOf timestamp
    ```
   // ensure tombstone is skipped after an AOST skip
   {input: "b3c2Xc1d1", expectedNextKey: "d1", expectedNext: "d1", asOf: "2"},
   {input: "b3c2Xc1d1", expectedNextKey: "c1d1", expectedNext: "c1d1", asOf: "1"},
    ```

- the iterator ignores _all_ tombstones above the timestamp. See tests: 
    ```
    // ensure an AOST 'next' takes precedence over a tombstone 'nextkey'
   {input: "b2Xb1c1", expectedNextKey: "c1", expectedNext: "c1", asOf: ""},
   {input: "b2Xb1c1", expectedNextKey: "b1c1", expectedNext: "b1c1", asOf: "1"},
   ```

   and in the seek test
   ```
   // Ensure AOST 'next' takes precendence over tombstone 'nextkey'
  {"a4a3Xa1b1", "a3", "a1", "1"},
  {"a4a3Xa2a1b1", "a2", "a1", "1"},
  {"a4a3Xa2a1b1", "a2", "a2", "2"},
   ```

- the iterator will surface any keys that shadow a tombstone iff those keys are below the asOf timestamp. (I now realize I shouldn't use the word 'before' in this code comment). 
   ```
  // Ensure key before delete tombstone gets read if under AOST
  {input: "b2b1Xc1", expectedNextKey: "b2c1", expectedNext: "b2c1", asOf: ""},
  {input: "b2b1Xc1", expectedNextKey: "c1", expectedNext: "c1", asOf: "1"},
  ```
  
  To obey these invariants, the logic in `SeekGE` is a bit weird. Specifically, SeekGE needs to ensure we do _not_ return a key that's shadowed by a tombstone that's below the asOf timestamp.

Comment (erikgrinaker) [reply pkg/storage/read_as_of_iterator.go:15]: I see, thanks for clarifying. I'm curious why we don't care about old versions below a tombstone, but we do if they're below a visible key -- what's the use-case here? Naïvely, I would think we'd either not care about old versions at all, or we'd care about all of them.

>To obey these invariants, the logic in SeekGE is a bit weird, but the https://github.com/cockroachdb/cockroach/pull/77281#discussion_r873765645 you propose is correct. Specifically, SeekGE needs to ensure we do not return a key that's shadowed by a tombstone that's below the asOf timestamp.

Right, makes sense. So the simplification would be this one then (which potentially avoids a lot of wasted work)?

> this can be simplified by using the lowest timestamp value of `originalKey.Timestamp` and `f.asOf` and calling `advance()` once

Comment (msbutler) [reply pkg/storage/read_as_of_iterator.go:15]: >  I'm curious why we don't care about old versions below a tombstone, but we do if they're below a visible key -- what's the use-case here?

Yeah, this is an interesting point. I built this iterator to reflect the logic previously handled in `restore_processor.go`. This includes:
- Skipping to the next key if we see a tombstone below the asOf timestamp. I.e., we don't care about old versions below a tombstone.
   ```
   if len(iter.UnsafeValue()) == 0 {
			// Value is deleted.
			iter.NextKey()
			continue
		}
   ``` 
- Now, should we surface keys below a visible key? I think we should, even though the old restore processor logic didn't have any guidance on this. Currently, as soon as it finds a valid visible key, it calls `NextKey`. In other words, we _currently_ only restore at most  one version of  each key. But I could imagine a future version of Restore that requires restoring multiple versions of a key (eg https://github.com/cockroachdb/cockroach/issues/76722, which will allow a restoring cluster to finish or cancel an import from a backup cluster). 

On your other question:
> So the simplification would be this one then (which potentially avoids a lot of wasted work)?

Yes, exactly!

Comment (erikgrinaker) [reply pkg/storage/read_as_of_iterator.go:15]: >I could imagine a future version of Restore that requires restoring multiple versions of a ke

In these cases, wouldn't we want to also restore versions below tombstones? Otherwise, we'd have an incomplete history.

Anyway, I'll defer to @dt on this. I don't really have a strong opinion on these semantics as long as they cover our needs and they're consistent and well-documented.

Comment (msbutler) [reply pkg/storage/read_as_of_iterator.go:15]: just spoke with @dt about this. He doesn't actually think RESTORE should restore multiple versions of the same key. (my understanding of the import issue I tagged was incorrect). So for the purposes of restore, the `readAsOfIterator` should only ever surface one valid version per key, implying `Next` and `NextKey` should have the same implementation.

File: pkg/ccl/backupccl/restore_data_processor.go
Comment (msbutler) [reply pkg/ccl/backupccl/restore_data_processor.go:408]: Done.

File: pkg/storage/read_as_of_iterator.go
Comment (msbutler) [reply pkg/storage/read_as_of_iterator.go:19]: Good point. The iterator does not need to handle write intents, since it's currently only used by restore, which does not encounter any write intents.

File: pkg/storage/read_as_of_iterator.go
Comment (msbutler) [reply pkg/storage/read_as_of_iterator.go:29]: Done.

File: pkg/storage/read_as_of_iterator.go
Comment (msbutler) [reply pkg/storage/read_as_of_iterator.go:41]: Yes, good catch. Keys above the asOf timestamp should not matter. We should instead initially seek to the following key: `synthetic := MVCCKey{Key: originalKey.Key, Timestamp: f.asOf}`

File: pkg/storage/read_as_of_iterator.go
Comment (msbutler) [reply pkg/storage/read_as_of_iterator.go:91]: cheers!

File: pkg/storage/read_as_of_iterator_test.go
Comment (msbutler) [reply pkg/storage/read_as_of_iterator_test.go:116]: ah. good catch. I was blindly copying code from `multi_iterator_test.go`.

File: pkg/ccl/backupccl/restore_data_processor.go
Comment (erikgrinaker) [reply pkg/ccl/backupccl/restore_data_processor.go:408]: Doesn't appear to be?

File: pkg/storage/read_as_of_iterator.go
Comment (erikgrinaker) [reply pkg/storage/read_as_of_iterator.go:19]: Cool, let's add a mention of that in this comment then.

File: pkg/ccl/backupccl/restore_data_processor.go
Comment (msbutler) [reply pkg/ccl/backupccl/restore_data_processor.go]: done.

File: pkg/storage/read_as_of_iterator.go
Comment (msbutler) [reply pkg/storage/read_as_of_iterator.go]: Done.

File: pkg/storage/read_as_of_iterator.go
Comment (erikgrinaker) [inline pkg/storage/read_as_of_iterator.go:43]: nit: this check isn't necessary, since an empty timestamp sorts before all other timestamps in MVCC keys -- passing `f.asOf` will do the right thing.

**Review by erikgrinaker — APPROVED** [review]

File: pkg/storage/read_as_of_iterator.go
Comment (erikgrinaker) [reply pkg/storage/read_as_of_iterator.go:49]: Sorry, my bad -- just tried `TestReadAsOfIterator`, it's covered by `TestReadAsOfIteratorSeek`.

**Comment by erikgrinaker** [comment]
PS: remember to squash the commits before merging.

File: pkg/storage/read_as_of_iterator.go
Comment (msbutler) [reply pkg/storage/read_as_of_iterator.go:43]: huh, this is kinda subtle. When you pass a key with an empty timestamp to the underlying iterator's SeekGE, you're right, it will seek to the latest version of that key. So yes, we can remove this check, but only because we call the internal seek before doing any key comparisons. 

We can't, for example, remove this `isEmpty()` check in `advance` because an empty `hlcTimestamp.walltime` defaults to 0. 
```
} else if !f.asOf.IsEmpty() && f.asOf.Less(f.iter.UnsafeKey().Timestamp) {
```

Comment (erikgrinaker) [reply pkg/storage/read_as_of_iterator.go:43]: Yeah, `MVCCKey` and `Timestamp` order timestamps differently. The former orders them in descending order but with 0 first (which means that 3 is "less than" or "ordered before" 2, but 0 is before 3), while `Timestamp` uses regular ascending order (0,1,2,3...). I suppose you get used to it, but definitely subtle and a bit of a footgun.

**Pull request merged**
